src/nupic/algorithms/Connections.cpp

```diff
@@ -442,23 +444,26 @@ void Connections::computeActivity(
 void Connections::adaptSegment(const Segment segment, 
                                const SDR &inputs,
                                const Permanence increment,
-                               const Permanence decrement)
+                               const Permanence decrement,
+                               const vector<Permanence> &previousUpdates,
+                                     vector<Permanence> &currentUpdates)
 {
-  const vector<Synapse> &synapses = synapsesForSegment(segment);
-
   const auto &inputArray = inputs.getDense();
 
-  for (SynapseIdx i = 0; i < synapses.size(); i++) {
-    const SynapseData &synapseData = dataForSynapse(synapses[i]);
+  for( const auto &synapse : synapsesForSegment(segment) ) {
+    const SynapseData &synapseData = dataForSynapse(synapse);
 
-    Permanence permanence = synapseData.permanence;
+    Permanence update;
     if( inputArray[synapseData.presynapticCell] ) {
-      permanence += increment;
+      update = increment;
     } else {
-      permanence -= decrement;
+      update = -decrement;
     }
 
-    updateSynapsePermanence(synapses[i], permanence);
+    if( update != previousUpdates[synapse] ) {
+      updateSynapsePermanence(synapse, synapseData.permanence + update);
+    }
+    currentUpdates[ synapse ] = update;
```

Comment:
what's the size of currentUpdates? Rather push_back than [idx]=x? 
Also, can you preallocate currentUpdates to some sensible value?

Reply:
Both of these vectors will be resized to "con.synapseFlatListLength()"  If the users gives this method an empty vector it will be correctly resized & filled with the correct value (zero).